src/ui/public/state_management/unhash_states.js

```diff
@@ -0,0 +1,43 @@
+import { parse as parseUrl, format as formatUrl } from 'url';
+import { mapValues } from 'lodash';
+
+export function UnhashStatesProvider(getAppState, globalState) {
+  const getDefaultStates = () => [getAppState(), globalState].filter(Boolean);
+
+  this.inAbsUrl = (urlWithHashes, states = getDefaultStates()) => {
+    if (!urlWithHashes) return urlWithHashes;
+
+    const urlWithHashesParsed = parseUrl(urlWithHashes, true);
+    if (!urlWithHashesParsed.hostname) {
+      // passing a url like "localhost:5601" or "/app/kibana" should be prevented
+      throw new TypeError(
+        'Only absolute urls should be passed to `unhashStates.inAbsUrl()`. ' +
+        'Unable to detect url hostname.'
+      );
+    }
+
+    if (!urlWithHashesParsed.hash) return urlWithHashes;
+
+    const appUrl = urlWithHashesParsed.hash.slice(1); // trim the #
+    if (!appUrl) return urlWithHashes;
+
+    const appUrlParsed = parseUrl(urlWithHashesParsed.hash.slice(1), true);
+    if (!appUrlParsed.query) return urlWithHashes;
+
+    const appQueryWithoutHashes = this.inParsedQueryString(appUrlParsed.query || {}, states);
+    return formatUrl({
+      ...urlWithHashesParsed,
+      hash: formatUrl({
+        pathname: appUrlParsed.pathname,
+        query: appQueryWithoutHashes,
+      })
+    });
+  };
+
+  this.inParsedQueryString = (parsedQueryString, states = getDefaultStates()) => {
+    return mapValues(parsedQueryString, (val, key) => {
+      const state = states.find(s => key === s.getQueryParamName());
+      return state ? state.translateHashToRison(val) : val;
+    });
+  };
+}
```

Comment:
It seems like `getDefaultStates` is a bit of an implicit dependency, and I wonder if the complexity it adds is necessary? What if we extracted it into its own service called something like `get_hashed_url_states.js`, and removed the `states` parameter default value (i.e. no longer assign `states = getDefaultStates()`)?

Then we would gain a few things:

1. This file would have fewer implicit dependencies.
2. It could also then be split up into separate files, e.g. `unhash_url.js` and `unhash_query_string.js` (I think these names are clearer, and show the cohesion between the two functions).
3. Everywhere we call `unhashUrl(hashedUrl, hashedUrlStates)` becomes a little more transparent, because you can read that and infer the outcome of calling this method in relationship to what we're passing into it.